### ISSUE TITLE
mission.cpp: No matching function for call to 'find_if'

### ISSUE THREAD

**Issue by kreatordxx** [dxx-rebirth/dxx-rebirth#31]
Got another one here:
> similar/main/mission.cpp:881:15: error: no matching function for call to
>       'find_if'
>                                         auto s = std::find_if(lb, t, a);
>                                                  ^~~~~~~~~~~~
> /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:856:1: note: 
>       candidate template ignored: deduced conflicting types for parameter
>       '_InputIterator' ('const char *' vs. 'char *')
> find_if(_InputIterator __first, _InputIterator __last, _Predicate __pred)
> ^
> 1 error generated.

**Comment by vLKp** [comment]
It looks like your strchr follows the old C convention of stripping **const** from the input.  As I understand it, this was fixed in C++98.  I can work around it easily enough, but before I do, please check whether this patch fixes that call site.
```diff
--- a/similar/main/mission.cpp
+++ b/similar/main/mission.cpp
@@ -27,7 +27,7 @@ COPYRIGHT 1993-1999 PARALLAX SOFTWARE CORPORATION.  ALL RIGHTS RESERVED.
 #include <vector>
 #include <stdio.h>
 #include <stdlib.h>
-#include <string.h>
+#include <cstring>
 #include <ctype.h>
 #include <limits.h>
 
```

If it does not, please try also:
```diff
--- a/similar/main/mission.cpp
+++ b/similar/main/mission.cpp
@@ -872,7 +872,7 @@ static int load_mission(const mle *mission)
 						break;
 					const auto &line = buf.line();
 					const auto lb = line.begin();
-					const auto t = strchr(lb, ',');
+					const auto t = std::strchr(lb, ',');
 					if (!t)
 						break;
 					auto a = [](char c) {
```

**Comment by btb** [comment]
I'm seeing the same issue, and neither of those have any effect.
this does compile:
```
diff --git a/similar/main/mission.cpp b/similar/main/mission.cpp
index 6962ef3..de19831 100644
--- a/similar/main/mission.cpp
+++ b/similar/main/mission.cpp
@@ -923,7 +923,7 @@ static int load_mission(const mle *mission)
                                        if (!PHYSFSX_fgets(buf, mfile))
                                                break;
                                        const auto &line = buf.line();
-                                       const auto lb = line.begin();
+                                       const auto lb = (char *)line.begin();
                                        const auto t = strchr(lb, ',');
                                        if (!t)
                                                break;
```

**Comment by kreatordxx** [comment]
I can confirm exactly the same as Bradley.

**Comment by vLKp** [comment]
The mingw32 headers also get this wrong, so I can reproduce this with a cross-compile to mingw32.

Casting away const-ness is almost always wrong, which is why `strchr` is wrong to strip it.  The fix is either to cast the result of `strchr` to const or to make `t` have type `const char *` instead of `auto`.

Fixed in ea3c789f646a07ebb4120ebe8c969843fdbb5850 by giving `t` type `const char *`.

**Comment by kreatordxx** [comment]
Excellent, I can confirm it now builds and links with SCons and Clang 6.0.
